File: solutions/code-carl/ch04_string/T4_LC0028_StrStr.java
/**
 * Practice template for:
 * - LeetCode 28. Find the Index of the First Occurrence in a String
 *   https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/
 */
public class T4_LC0028_StrStr {
    public int strStr(String haystack, String needle) {
        char[] string = haystack.toCharArray();
        char[] patt = needle.toCharArray();
        int[] next = new int[patt.length];
        int j;

        j = 0;
        for (int i = 1; i < next.length; i++) {
            while (j > 0 && patt[i] != patt[j]) {
                j = next[j - 1];
            }
            if (patt[i] == patt[j]) {
                j++;
            }
            next[i] = j;
        }

        j = 0;
        for (int i = 0; i < string.length; i++) {
            while (j > 0 && string[i] != patt[j]) {
                j = next[j - 1];
            }
            if (string[i] == patt[j]) {
                j++;
                if (j == patt.length) {
                    return i - j + 1;
                }
            }
        }
        return -1;
    }
/*    public int strStr(String haystack, String needle) {
        char[] string = haystack.toCharArray();
        char[] patt = needle.toCharArray();
        int[] next = new int[patt.length];
        next[0] = 0;
        int i, j;

        j = 0;
        for (i = 1 ;i < patt.length; i++) {
            while (j > 0 && patt[i] != patt[j])
               j =  next[j - 1];
            if (patt[i] == patt[j])
                j++;
            next[i] = j;
        }

        j = 0;
        for (i  = 0; i < string.length; i ++) {
            while (j > 0 &&  string[i] != patt[j])
                j = next[j - 1];
            if (string[i] == patt[j])
                j++;
            if (j == patt.length)
                return i - j + 1;
        }
        return -1;
    }*/

}

File: solutions/python/code-carl/ch04_string/t4_lc0028_str_str.py
"""LeetCode 28. Find the Index of the First Occurrence in a String
https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/
"""
from typing import List


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        next: List[int] = [0] * len(needle)

        j = 0
        for i in range(1, len(needle)):
            while j > 0 and needle[i] != needle[j]:
                j = next[j - 1]
            if needle[i] == needle[j]:
                j += 1
            next[i] = j

        j = 0
        for i in range(len(haystack)):
            while j > 0 and haystack[i] != needle[j]:
                j = next[j - 1]
            if haystack[i] == needle[j]:
                j += 1
                if j == len(needle):
                    return i - j + 1

        return -1


if __name__ == "__main__":
    solution = Solution()
    assert solution.strStr("sadbutsad", "sad") == 0
    assert solution.strStr("leetcode", "leeto") == -1
    assert solution.strStr("mississippi", "issip") == 4
    print("All tests passed.")
